File: Arrays/Derived_Structures/Deque/Implementation/C++/circular_deque_array.cpp
#include <bits/stdc++.h>
using namespace std;

class CircularDeque {
	vector<int> arr;
	int front, rear, size, capacity;
public:
	CircularDeque(int cap) : arr(cap), front(0), rear(-1), size(0), capacity(cap) {}
	void push_front(int x) {
		if (size == capacity) return;
		front = (front - 1 + capacity) % capacity;
		arr[front] = x;
		size++;
	}
	void push_back(int x) {
		if (size == capacity) return;
		rear = (rear + 1) % capacity;
		arr[rear] = x;
		size++;
	}
	void pop_front() {
		if (size == 0) return;
		front = (front + 1) % capacity;
		size--;
	}
	void pop_back() {
		if (size == 0) return;
		rear = (rear - 1 + capacity) % capacity;
		size--;
	}
	int getFront() { return size == 0 ? -1 : arr[front]; }
	int getRear() { return size == 0 ? -1 : arr[rear]; }
	bool empty() { return size == 0; }
};

int main() {
	CircularDeque d(3);
	d.push_back(1); d.push_front(2); d.push_back(3);
	cout << d.getFront() << ' ' << d.getRear() << endl;
	d.pop_front();
	cout << d.getFront() << ' ' << d.getRear() << endl;
	cout << (d.empty() ? "Empty" : "Not Empty") << endl;
	return 0;
}

File: Arrays/Derived_Structures/Deque/Implementation/Python/circular_deque_array.py
class CircularDeque:
	def __init__(self, cap):
		self.arr = [None] * cap
		self.front = 0
		self.rear = -1
		self.size = 0
		self.capacity = cap
	def push_front(self, x):
		if self.size == self.capacity:
			return
		self.front = (self.front - 1 + self.capacity) % self.capacity
		self.arr[self.front] = x
		self.size += 1
	def push_back(self, x):
		if self.size == self.capacity:
			return
		self.rear = (self.rear + 1) % self.capacity
		self.arr[self.rear] = x
		self.size += 1
	def pop_front(self):
		if self.size == 0:
			return
		self.front = (self.front + 1) % self.capacity
		self.size -= 1
	def pop_back(self):
		if self.size == 0:
			return
		self.rear = (self.rear - 1 + self.capacity) % self.capacity
		self.size -= 1
	def get_front(self):
		return self.arr[self.front] if self.size != 0 else None
	def get_rear(self):
		return self.arr[self.rear] if self.size != 0 else None
	def empty(self):
		return self.size == 0

if __name__ == "__main__":
	d = CircularDeque(3)
	d.push_back(1); d.push_front(2); d.push_back(3)
	print(d.get_front(), d.get_rear())
	d.pop_front()
	print(d.get_front(), d.get_rear())
	print("Empty" if d.empty() else "Not Empty")
class CircularDeque:
	def __init__(self, cap):
		self.arr = [None] * cap
		self.front = 0
		self.rear = -1
		self.size = 0
		self.capacity = cap
	def push_front(self, x):
		if self.size == self.capacity:
			return
		self.front = (self.front - 1 + self.capacity) % self.capacity
		self.arr[self.front] = x
		self.size += 1
	def push_back(self, x):
		if self.size == self.capacity:
			return
		self.rear = (self.rear + 1) % self.capacity
		self.arr[self.rear] = x
		self.size += 1
	def pop_front(self):
		if self.size == 0:
			return
		self.front = (self.front + 1) % self.capacity
		self.size -= 1
	def pop_back(self):
		if self.size == 0:
			return
		self.rear = (self.rear - 1 + self.capacity) % self.capacity
		self.size -= 1
	def get_front(self):
		return self.arr[self.front] if self.size != 0 else None
	def get_rear(self):
		return self.arr[self.rear] if self.size != 0 else None
	def empty(self):
		return self.size == 0

if __name__ == "__main__":
	d = CircularDeque(3)
	d.push_back(1); d.push_front(2); d.push_back(3)
	print(d.get_front(), d.get_rear())
	d.pop_front()
	print(d.get_front(), d.get_rear())
	print("Empty" if d.empty() else "Not Empty")
